docs/constrained-generators/manual.md

```diff
@@ -324,125 +568,1080 @@ match
+
+## Strategy 2 writing your own SimpleRep instance
+
+This strategy depends on the provided `(SimpleRep T)` associated type family instance, being an actual Sum-of-Products type.
+This requires quite a bit of knowledge about the internals of the system. Let's look closely at the `SimpleRep` class
+
+```
+class Typeable (SimpleRep a) => HasSimpleRep a where
+  type SimpleRep :: * -> *
+  type family SimpleRep a
+  toSimpleRep :: a -> SimpleRep a
+  fromSimpleRep :: HasSimpleRep a => SimpleRep a -> a
+```
+
+What kind of type lends itself to this strategy? 
+1.  A type that has internal structure that enforces some internal invariants.
+2.  A type that has a -builder- function, that takes simple input, and constructs the internal struture.
+3.  A type that has an -accessor- function, that takes the internal structure, and returns the simple input.
+4.  A type where the -simple input- has a Sum-of-Products representation.
+
+Often the -builder- function is implemented as a Haskell Pattern.  Here is an example that come from the Cardano Ledger.
+A lot of complicated stuff is not fully describe here, but the example gives an overview of how it works.
+
+
+```
+-- NOTE: this is a representation of the `ShelleyTxOut` type. You can't
+-- simply use the generics to derive the `SimpleRep` for `ShelleyTxOut`
+-- because the type is memoized (i.e. it stores a hidden copy of the actual bytes that
+-- were deserialized when the TxOut was transmitted over the network). So instead,
+-- we say that the representation is the same as what you would get from 
+-- using the `ShelleyTxOut` pattern.
+type ShelleyTxOutTypes era =
+  '[ Addr
+   , Value era
+   ]
+instance (Era era, Val (Value era)) => HasSimpleRep (ShelleyTxOut era) where
+  type TheSop (ShelleyTxOut era) = '["ShelleyTxOut" ::: ShelleyTxOutTypes era]
+  toSimpleRep (ShelleyTxOut addr val) =
+    inject @"ShelleyTxOut" @'["ShelleyTxOut" ::: ShelleyTxOutTypes era]
+      addr
+      val
+  fromSimpleRep rep =
+    algebra @'["ShelleyTxOut" ::: ShelleyTxOutTypes era] rep ShelleyTxOut
+
+instance (EraTxOut era, HasSpec (Value era)) => HasSpec (ShelleyTxOut era)
+```
+
+TODO add more explanation about the types. Much of the example of depends on properties of TxOut
+that is not explained.
+
+## Strategy 3 defining the SimpleRep instance in terms of another type with a SimpleRep instance
+
+The type `Coin` is a defined as
+
+```
+newtype Coin = Coin {unCoin :: Integer}
+```
+The operations on Coin ensure the invariant that one cannot build a `Coin` with a negative value.
+We can enforce these invariants in a `SimpleRep` instance by making the `SimpleRep` type be a `Word64`.
+`Word64` is one of the numeric types, and has its own `SimpleRep` instance, so it is a good choice.
+Here is the `SimpleRep` instance, and the `HasSpec` instance defined using that representation.
+
+```
+instance HasSimpleRep Coin where
+  type SimpleRep Coin = Word64
```

Comment:
Maybe use `Natural` instead of `Word64` here just because that's morally what `Coin` _really_ is?